blog/content/post/2022-05-04-fobnail_provisioning.md

```diff
@@ -0,0 +1,302 @@
+it should follow.
+
+First resource that Fobnail Token asks for during its provisioning is a Platform
+Owner certificate chain. It has a [strictly defined format](TBD: link to docs)
+and its root must be provided to the token during compilation. Fobnail validates
+this chain, but because it has neither timekeeping nor (secure) networking
+capabilities, it isn't able to check validity periods or CRLs.
+
+Second request uses POST method and includes CSR (certificate signing request).
+Platform Owner responses to that with a certificate generated from that CSR,
+signed by Platform Owner's private key. This certificate is verified against
+chain received in previous step and saved in non-volatile memory for later use.
+
+![Diagram of Fobnail provisioning flow](/img/Fobnail-flows-fobnail-provisioning.png)
+
+## Fobnail Token's key
+
+There are no good crates for generating RSA keys on Cortex-M, and in our tests
+using universal implementation wasn't able to generate 2048-bit key in 5 hours.
+We also couldn't use ARM TrustZone CryptoCell 310 which is included in nRF52840
+for this purpose because it doesn't have open-sourced libraries. Thus, currently
+Fobnail Token generates Curve25519 key, with help of [Trussed](https://trussed.dev/).
+
+It was originally designed to be used in Diffie–Hellman method to generate
+symmetric encryption key for communication between owners of two separate
+Curve25519 keys. Ed25519 is a signature scheme using Curve25519, which adds the
+ability to sign and verify data, including certificates. Unfortunately,
+Curve25519 can't be used as encryption key pair. This is a problem that will
+have to be resolved to unlock full power of Fobnail Token.
+
+## Possible uses of Fobnail certificate
+
+Fobnail architecture doesn't impose use of key for which certificate is being
+made. The only restriction that should be followed is that created certificate
+is not a CA, because that together with inability to use CRLs creates enormous
+security issue.
+
+Usage of key should answer the question: "why do I want to attest the platform"?
+There is no universal answer, each use case has its own requirements. Example
+usages with corresponding X.509 `keyUsage` bits include, but are not limited to:
+
+- obtaining small data (e.g. nonce, password, salt) from encrypted challenge,
+  for this `dataEncipherment` must be set (note: can't be done with Curve25519),
+- authentication, in which case `digitalSignature` and `nonRepudiation` is good
+  enough,
+- handshake in TLS-like communication, which requires `keyAgreement` or
+  `keyEncipherment`.
+
+In any case, private key must never leave Fobnail Token. This may limit speed
+and maximal size of data that is signed or encrypted, as those tasks must be
+performed by the Token. For example, encrypting whole hard disk with Fobnail's
+public key directly is a bad idea, but you can use it to encrypt a key used for
+decrypting the drive.
```

Comment:
@krystian-hebel what may mean potential LUKS intergation?

Reply:
That is one of possible use cases, yes.